### PR TITLE
HPCC-26092 Tighten up default plane validation

### PR DESCRIPTION
**Pull request by richardkchapman**
Signed-off-by: Richard Chapman <rchapman@hpccsystems.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [x] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

File: dockerfiles/startall.sh [hunk 1]
```diff
@@ -95,10 +105,11 @@ if [[ -n ${PERSIST} ]] ; then
   mkdir -p ${PERSIST_PATH}/hpcc-data
```
Comment (ghalliday) [inline dockerfiles/startall.sh:105]: minor: could also execute this if -n ${PERSISTVALUES}

File: helm/hpcc/templates/_helpers.tpl [hunk 2]
```diff
@@ -514,15 +520,11 @@ Check that the storage and spill planes for a component exist
 {{- define "hpcc.checkDefaultStoragePlane" -}}
 {{- if (hasKey .me "storagePlane") }}
  {{- $search := .me.storagePlane -}}
- {{- if ne (include "hpcc.isValidStoragePlane" (dict "search" $search "root" .root)) "true" -}}
-  {{- $_ := fail (printf "storage data plane %s for %s is not defined" $search .me.name ) }}
- {{- end -}}
+ {{- include "hpcc.isValidStoragePlane" (dict "search" $search "root" .root "category" "data" "type" "storage data" "for" .me.name) -}}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:523]: minor, better to rename isValid to checkValid

Comment (ghalliday) [reply helm/hpcc/templates/_helpers.tpl:523]: Alternative implementation would be to have a function that returns a list of all the storage planes, and include it in the fail message.

File: helm/hpcc/templates/_helpers.tpl [hunk 3]
```diff
@@ -491,21 +491,27 @@ vaults:
...
 {{- $storage := (.root.Values.storage | default dict) -}}
 {{- $planes := ($storage.planes | default list) -}}
 {{- $dataStorage := ($storage.dataStorage | default dict) -}}
 {{- /* If storage.dataStorage.plane is defined, the implicit plane hpcc-dataplane is not defined */ -}}
-{{- $done := dict "matched" (and (not $dataStorage.plane) (eq $search "hpcc-dataplane")) -}}
+{{- $done := dict "matched" (and (not $dataStorage.plane) (eq $search "hpcc-dataplane")) "all" "" -}}
 {{- range $plane := $planes -}}
- {{- if eq $search $plane.name -}}
- {{- $_ := set $done "matched" true -}}
+ {{- if has $category $plane.labels -}}
```
Comment (ghalliday) [inline helm/hpcc/templates/_helpers.tpl:505]: better to reverse the tests, and report an error if the category isn't in the labels.  Also need to be careful at the moment that a blank labels list is treated as equivalent to [data]

**Review by ghalliday — COMMENTED** [review]
@richardkchapman some minor comments

File: helm/hpcc/templates/_helpers.tpl (hunk 3 above)
Comment (richardkchapman) [reply helm/hpcc/templates/_helpers.tpl:505]: Except I'm scanning through all planes with this category in order to give the candidate list in the error message

File: dockerfiles/startall.sh (hunk 1 above)
Comment (richardkchapman) [reply dockerfiles/startall.sh:105]: Should really be creating directories that are named in the values file in that case

**Comment by richardkchapman** [comment]
@ghalliday I don't know if this is still useful given other changes

**Comment by ghalliday** [comment]
I think it probably is.  There probably need to be more tests like this, but this is a useful start.

**Review by ghalliday — APPROVED** [review]

**Pull request merged**